synapseclient/api/file_services.py

```diff
@@ -0,0 +1,335 @@
+        synapse_client: If not passed in or None this will use the last client from
+            the `.login()` method.
+
+    Returns:
+        A FileHandle for objects that are stored externally.
+        <https://rest-docs.synapse.org/rest/org/sagebionetworks/repo/model/file/ExternalFileHandleInterface.html>
+    """
+    from synapseclient import Synapse
+
+    client = Synapse.get_client(synapse_client=synapse_client)
+    file_name = external_url.split("/")[-1]
+    external_url = utils.as_url(external_url)
+    file_handle = {
+        "concreteType": concrete_types.EXTERNAL_FILE_HANDLE,
+        "fileName": file_name,
+        "externalURL": external_url,
+        "contentMd5": md5,
+        "contentSize": file_size,
+    }
+    if mimetype is None:
+        (mimetype, _) = mimetypes.guess_type(external_url, strict=False)
+    if mimetype is not None:
+        file_handle["contentType"] = mimetype
+    return await client.rest_post_async(
+        "/externalFileHandle", json.dumps(file_handle), client.fileHandleEndpoint
+    )
+
+
+async def post_external_s3_file_handle(
+    bucket_name: str,
+    s3_file_key: str,
+    file_path: str,
+    parent: str = None,
+    storage_location_id: str = None,
+    mimetype: str = None,
+    md5: str = None,
+    synapse_client: Optional["Synapse"] = None,
+) -> Dict[str, Union[str, int, bool]]:
+    """
+    Create an external S3 file handle for e.g. a file that has been uploaded directly to
+    an external S3 storage location.
+
+    <https://rest-docs.synapse.org/rest/POST/externalFileHandle/s3.html>
+
+    Arguments:
+        bucket_name: Name of the S3 bucket
+        s3_file_key: S3 key of the uploaded object
+        file_path: Local path of the uploaded file
+        parent: Parent entity to create the file handle in, the file handle will be
+            created in the default storage location of the parent. Mutually exclusive
+            with storage_location_id
+        storage_location_id: Explicit storage location id to create the file handle in,
+            mutually exclusive with parent
+        mimetype: Mimetype of the file, if known
+        md5: MD5 of the file, if known
+        synapse_client: If not passed in or None this will use the last client from
+            the `.login()` method.
+
+    Returns:
+        The created file handle.
+        <https://rest-docs.synapse.org/rest/org/sagebionetworks/repo/model/file/S3FileHandle.html>
+
+    Raises:
+        ValueError: If neither parent nor storage_location_id is specified, or if
+            both are specified.
+    """
+    from synapseclient import Synapse
+
+    client = Synapse.get_client(synapse_client=synapse_client)
+
+    if storage_location_id:
+        if parent:
+            raise ValueError("Pass parent or storage_location_id, not both")
+    elif not parent:
+        raise ValueError("One of parent or storage_location_id is required")
+    else:
+        upload_destination = await get_upload_destination(
```

Comment:
I guess not related to this PR, but was this func meant to be named `get_upload_destination_async`?

Reply:
I am not planning to include this for all the api functions. They're geared towards internal use and I am not providing both a sync and async call for them.